src/mpi/coll/reduce/reduce_intra_reduce_scatter_gather_const_space.c

```diff
@@ -0,0 +1,387 @@
+   total cost is:
+   Cost = 2.lgp.alpha + 2.n.((p-1)/p).beta + n.((p-1)/p).gamma
+
+   For the non-power-of-two case, assuming the root is not one of the
+   odd-numbered processes that get excluded in the reduce-scatter,
+   Cost = (2.floor(lgp)+1).alpha + (2.((p-1)/p) + 1).n.beta + 
+           n.(1+(p-1)/p).gamma
+   
+*/
+#undef FUNCNAME
+#define FUNCNAME MPIR_Reduce_intra_reduce_scatter_gather_const_space
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_Reduce_intra_reduce_scatter_gather_const_space (
+    const void *sendbuf,
+    void *recvbuf,
+    int count,
+    MPI_Datatype datatype,
+    MPI_Op op,
+    int root,
+    MPIR_Comm *comm_ptr,
+    MPIR_Errflag_t *errflag )
+{
+    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno_ret = MPI_SUCCESS;
+    int comm_size, rank, type_size ATTRIBUTE((unused)), pof2, rem, newrank, rank_gather;
+    int mask, i, j, newdst;
+    int dst, newroot, newdst_tree_root, newroot_tree_root;
+    MPI_Aint true_lb, true_extent, extent;
+    void *tmp_buf;
+    int lsize, rsize, sdisp, rdisp, rcnt, scnt, iter, niter, newcount;
+
+    MPIR_CHKLMEM_DECL(2);
+
+    comm_size = comm_ptr->local_size;
+    rank = comm_ptr->rank;
+
+    /* set op_errno to 0. stored in perthread structure */
+    {
+        MPIR_Per_thread_t *per_thread = NULL;
+        int err = 0;
+
+        MPID_THREADPRIV_KEY_GET_ADDR(MPIR_ThreadInfo.isThreaded, MPIR_Per_thread_key,
+                                     MPIR_Per_thread, per_thread, &err);
+        MPIR_Assert(err == 0);
+        per_thread->op_errno = 0;
+    }
+
+    /* Create a temporary buffer */
+
+    MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+    MPIR_Datatype_get_extent_macro(datatype, extent);
+
+    /* I think this is the worse case, so we can avoid an assert() 
+     * inside the for loop */
+    /* should be buf+{this}? */
+    MPIR_Ensure_Aint_fits_in_pointer(count * MPL_MAX(extent, true_extent));
+
+    MPIR_CHKLMEM_MALLOC(tmp_buf, void *, count*(MPL_MAX(extent,true_extent)),
+                        mpi_errno, "temporary buffer", MPL_MEM_BUFFER);
+    /* adjust for potential negative lower bound in datatype */
+    tmp_buf = (void *)((char*)tmp_buf - true_lb);
+    
+    /* If I'm not the root, then my recvbuf may not be valid, therefore
+       I have to allocate a temporary one */
+    if (rank != root) {
+        MPIR_CHKLMEM_MALLOC(recvbuf, void *,
+                            count*(MPL_MAX(extent,true_extent)), 
+                            mpi_errno, "receive buffer", MPL_MEM_BUFFER);
+        recvbuf = (void *)((char*)recvbuf - true_lb);
+    }
+
+    if ((rank != root) || (sendbuf != MPI_IN_PLACE)) {
+        mpi_errno = MPIR_Localcopy(sendbuf, count, datatype, recvbuf,
+                                   count, datatype);
+        if (mpi_errno) { MPIR_ERR_POP(mpi_errno); }
+    }
+
+    MPIR_Datatype_get_size_macro(datatype, type_size);
+
+    /* find the number of steps in reduce scatter */
```

Comment:
This calculation of `pof2` is no longer needed given the assignment below. It should be removed.

Reply:
@Bayyapu be careful to accurately calculate the number of iterations when removing this.

Reply:
Ah, nevermind. The loop actually goes beyond `comm_size`, so needs to be reset back to the correct value. I think there is some room to go through all the collective code and cleanup these types of `pof2` and base 2 log calculations, but we can do that in a separate ticket, I think.